k2/python/csrc/torch/mutual_information_cuda.cu

```diff
@@ -0,0 +1,821 @@
+    if (threadIdx.x == 0) {
+      // This if-statement is an optimization and modification of the loop below
+      // for the value i == 0, i.e. inner-iteration == 0.  The modification is
+      // to set p_buf to 1.0 = exp(0.0) if this is the "origin block",
+      // i.e. s == s_begin, t == t_begin.  This corresponds to the
+      // probability of the pair of sequences of length (0, 0).
+      p_buf[1][1] =
+          (is_origin_block ? 0.0
+                           : LogAdd<scalar_t>()(p_buf[0][1] + px_buf[0][0],
+                                                p_buf[1][0] + py_buf[0][0]));
+    }
+
+    scalar_t p_buf_s1_t;  // This is for an optimization to avoid one
+                          // shared-memory read/write in the loop below.  it
+                          // represents p_buf[s + 1][t]; the first time we
+                          // access this, it will be for t == 0, except for
+                          // thread 0 when we first need it for t == 1.
+    if (threadIdx.x < BLOCK_SIZE) {
+      int s = threadIdx.x;
+      p_buf_s1_t = p_buf[s + 1][threadIdx.x == 0 ? 1 : 0];
+    }
+
+    int s = threadIdx.x;
+    for (int i = 1; i < block_S + block_T - 1; ++i) {
+      __syncwarp();
+      // i is the inner iteration, which corresponds to the (s + t) indexes of
+      // the elements within the block that we write.  So i == 0 writes
+      // positions (s, t) == (0, 0) (but we treated i == 0 as a special case
+      // above); i == 1 writes (0, 1) and (1, 0); i == 2 writes (0, 2), (1, 1)
+      // and (2, 1); and so on.  Note: not many threads participate in this
+      // part, only up to BLOCK_SIZE at most.  Unfortunately we couldn't figure
+      // out a very meaningful way for more threads to do work, that looked like
+      // it would really spead things up.
+      // So this kernel does (2 * BLOCK_SIZE) iterations, which may seem a lot,
+      // but we do at least do the I/O in an efficient way and keep the
+      // inner loop simple and fast (e.g. no exp() or log()).
+      int t = i - s;
+      if (s < block_S &&
+          static_cast<unsigned int>(t) < static_cast<unsigned int>(block_T)) {
+        // p_buf is indexed by s + 1 and t + 1 because it has an extra initial
+        // row and column for context from previous blocks.  Taking into account
+        // the way these buffers relate to the tensors p, px and py,
+        // can be interpreted as follows,
+        // writing sbb for s_block_begin and tbb for t_block_begin:
+        //
+        //   p[b][s+sbb][t+tbb] = LogAdd(p[b][s+sbb-1][t+tbb] +
+        //   px[s+sbb-1][t+tbb],
+        //                               p[b][s+sbb][t+tbb-1] +
+        //                               py[s+sbb][t+tbb-1]
+        //
+        // where you can see that apart from the offsets of tbb and sbb, this is
+        // the same as the recursion defined for p in
+        // mutual_information.py:mutual_information_recursion(); and (eq. 0)
+        // above.
+#if 1
```

Comment:
Oh, this is something I forgot to test...
More ideally this would be #if 0, that path is slightly more optimized, but I have not tested it.

Reply:
Ok, I will test it.